_delb/nodes.py

```diff
@@ -3006,17 +3080,347 @@ def not_wrapper(node: NodeBase) -> bool:
     return not_wrapper
 
 
+# serializer
+
+
+class SerializerBase:
+    __slots__ = (
+        "align_attributes",
+        "buffer",
+        "indentation",
+        "_level",
+        "_namespaces",
+        "_prefixes",
+        "__text_wrapper",
+    )
+
+    def __init__(self, buffer: TextIO):
+        self.buffer = buffer
+
+    def __enter__(self) -> Self:
+        self._level = 0
+        self._prefixes: dict[Optional[str], str] = {}
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.buffer.flush()
+
+    def __collect_prefixes(self, root: TagNode):
+        for node in traverse_bf_ltr_ttb(root, is_tag_node):
+            assert isinstance(node, TagNode)
+            for namespace in {node.namespace} | {
+                a.namespace for a in node.attributes.values()
+            }:
+                if namespace in self._prefixes:
+                    continue
+
+                if namespace is None:
+                    if "" in self._prefixes.values():
+                        # hence there's a default namespace in use
+                        self.__new_namespace_declaration(namespace)
+                    else:
+                        self._prefixes[None] = ""
+                    continue
+
+                try:
+                    self._namespaces._fallback = node.namespaces
+                    prefix = self._namespaces.lookup_prefix(namespace)
+
+                    if prefix is None:
+                        if "" in self._prefixes.values():
+                            # hence there's an empty namespace in use
+                            self._prefixes[namespace] = ""
+                            namespace = None
+                            raise KeyError
+                        else:
+                            prefix = ""
+                    else:
+                        prefix = prefix + ":"
+
+                    self._prefixes[namespace] = prefix
+
+                except KeyError:
+                    assert isinstance(namespace, str)
+                    self.__new_namespace_declaration(namespace)
+
+    def _init_config(
+        self,
+        align_attributes: bool,
+        indentation: str,
+        namespaces: None | NamespaceDeclarations,
+        text_width: int,
+    ):
+        self.align_attributes = align_attributes
+        self.indentation = indentation
+
+        if namespaces is None:
+            self._namespaces = Namespaces({})
+        else:
+            self._namespaces = Namespaces(namespaces)
```

Comment:
```suggestion
        self._namespaces = Namespaces(namespaces or {})
```
?

Reply:
no, b/c an empty `Namespaces` instance would also evaluate as `False` while it could have a `_fallback` that would then be lost. actually that circumstance shouldn't manifest, but on would need to be aware of this assumption / detail on another page.

also, when a `Namespace` instance is created with an instance, the `_fallback` is lost for sure.

Reply:
i hate these snipped views. the `_fallback` is irrelevant in this context.

Reply:
i can't commit the suggestion from this view. please commit and push how it fits you.